File: daily/105-Construct-binary-tree-from-preorder-and-inorder-traversal.go
package daily

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}

	rootVal := preorder[0]
	node := &TreeNode{Val: rootVal}

	// Find rootVal's position in inorder
	mid := 0
	for i, v := range inorder {
		if v == rootVal {
			mid = i
			break
		}
	}

	leftInorder := inorder[:mid]
	rightInorder := inorder[mid+1:]

	leftSize := len(leftInorder)
	leftPreorder := preorder[1 : 1+leftSize]
	rightPreorder := preorder[1+leftSize:]

	node.Left = buildTree(leftPreorder, leftInorder)
	node.Right = buildTree(rightPreorder, rightInorder)

	return node
}

File: daily/Binary Tree/105-Construct-binary-tree-from-preorder-and-inorder-traversal.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None

        rootVal = preorder[0]
        node = TreeNode(rootVal)

        mid = inorder.index(rootVal)
        left_inorder = inorder[:mid]
        right_inorder = inorder[mid + 1: ]

        left_preorder_size = len(left_inorder)
        left_preorder = preorder[1:1+left_preorder_size]
        right_preorder = preorder[left_preorder_size + 1: ]

        node.left = self.buildTree(left_preorder, left_inorder)
        node.right = self.buildTree(right_preorder, right_inorder)

        return node